noresm_v7_cam6_3_123: Add oslo-aero as an new aerosol scheme

mvertens opened the pull request:
Addition of oslo-aero as new aerosol scheme

Summary: Addition of oslo-aero as a new aerosol scheme via primarily the addition of new cpp-ifdefs. 

Contributors: @mvertens @ingvisau @gold2718 

Reviewers: @gold2718 

Purpose of changes: #97: Add oslo-aero as new aerosol scheme

Github PR URL: https://github.com/NorESMhub/CAM/pull/98

Changes made to build system: Changes for turning on oslo-aero 

Changes made to the namelist: new oslo-aero namelists

Changes to the defaults for the boundary datasets: oslo-aero boundary datasets

Substantial timing or memory changes: to be determined

The main aim of this PR is to being carrying out simulations with oslo-aero as the aerosol scheme used in CAM.  New OSLO_AERO CPP-ifdefs have been introduced that enable oslo-aero to be activated. This will be the first step for future object-oriented refactorization that is planned. 

Testing: NOTE: Regression testing of aux_cam_noresm completed.  
- Tests added for Oslo Aero
- Substantial changes were needed to ensure restart and BfB behavior when channging task count. This makes it impossible to compare with the release version of Oslo Aero.

AEROCOM is not enabled in this PR 

closes #89 
closes #97

Commit:
Restore pp_trop_mam_oslo from cam_cesm2_1_rel_05-Nor_v1.0.5

Copy pp_trop_mam_oslo mechanism and pp source from cam_cesm2_1_rel_05-Nor_v1.0.5
Add trop_mam_oslo chemistry option
Specify trop_mam_oslo for NF compsets
Set correct paths in configure
Some cleanup in configure
Bring in OSLO_AERO as a CAM external.

NB: Fortran is identical to that created with the new chemistry preprocessor

Commit:
Compile Oslo Aero from source subdirectories

mvertens committed:
fixes of run time errors

mvertens committed:
more updates to fix errors in the run

mvertens committed:
changed Externals_CAM

mvertens committed:
simplified usage of OSLO_AERO

mvertens committed:
added logic for getting dms flux from ocean to nuopc cap

mvertens committed:
removed trailing whitespace

mvertens committed:
updated testing and compsets

mvertens committed:
cleanup of oslo_aero differences

mvertens committed:
more simplifications for oslo_aero

mvertens committed:
more cleanup of radiation code

mvertens committed:
updates for compset

mvertens committed:
back out changes to be consistent with release code implementation

mvertens committed:
added new compsets

mvertens committed:
test updates

gold2718 commented on bld/build-namelist:
```diff
@@ -928,7 +933,15 @@ else {
 if ($prescribed_aero_model eq 'modal' or $chem =~ /_mam/) {$aer_model = 'mam';}
 
-if ($aer_model eq 'mam' ) {
+if ($chem =~ /_mam_oslo/) {$aer_model = 'oslo';}
```
I would move this up before line 934 and make 934 an `elsif`.
That make it clear that `_mam_oslo` takes precedence over `_mam` or `modal`.

gold2718 commented on bld/build-namelist:
```diff
@@ -928,7 +933,15 @@ else {
 
-if ($aer_model eq 'mam' ) {
+if ($chem =~ /_mam_oslo/) {$aer_model = 'oslo';}
+
+   #do nothing here
+
+if ($aer_model eq 'oslo') {
+
+   #do nothing here
+
+} elsif ($aer_model eq 'mam' ) {
```
I'm not sure why this is needed right now. Why not just:
```suggestion
if ($aer_model eq 'mam' ) {
```

gold2718 commented on bld/build-namelist:
```diff
@@ -2594,10 +2608,96 @@ if (($chem eq 'trop_mam4') or ($chem eq 'waccm_sc_mam4') or ($chem eq 'ghg_mam4'
+		           'isoprene -> ' => 'isoprene_oslo_emis_file',
+                 );
+
+    my $first = 1;
```
Not really needed (see below).

gold2718 commented on bld/build-namelist:
```diff
@@ -2594,10 +2608,96 @@ if (($chem eq 'trop_mam4') or ($chem eq 'waccm_sc_mam4') or ($chem eq 'ghg_mam4'
+        my $rel_filepath = get_default_value($species{$id} );
+        my $abs_filepath = set_abs_filepath($rel_filepath, $inputdata_rootdir);
+        $val .= $pre . quote_string($id . $abs_filepath);
+        if ($first) {
+            $pre = ",";
+            $first = 0;
+        }
```
This should work just as well as:
```suggestion
        $pre = ",";
```

gold2718 commented on bld/build-namelist:
```diff
@@ -2917,6 +3019,32 @@ $use_simple_phys = '.false.';
+ add_default($nl, 'ocean_filepath');
+ add_default($nl, 'dms_source');
+ add_default($nl, 'opom_source');
+#add_default($nl, 'dms_source_type');
+#add_default($nl, 'dms_cycle_year');
+#add_default($nl, 'ocean_filename');
+#add_default($nl, 'ocean_filepath');
+#add_default($nl, 'opom_source');
```
What's going on here?

gold2718 commented on bld/build-namelist:
```diff
@@ -2917,6 +3019,32 @@ $use_simple_phys = '.false.';
+    add_default($nl, 'dms_source_type');
+    add_default($nl, 'dms_cycle_year');
+}
+#add_default($nl, 'opom_source_type');
+#add_default($nl, 'opom_cycle_year');
```
What's going on here?

mvertens committed:
simplification of how radiation interacts with oslo_aero

mvertens committed:
fixed bug in radiation.F90

mvertens committed:
moved all files with OSLO_AERO ifdefs to src/chemistry/oslo_aero/src_cam and reset files to the original noresm_develop versions

gold2718 commented on cime_config/config_compsets.xml:
```diff
@@ -41,24 +41,53 @@
+    <lname>2000_CAM%DEV%NORESM%CAMoslo_CLM50%SP_CICE%PRES_DOCN%DOM_MOSART_SGLC_SWAV</lname>
+  </compset>
+  <compset>
+    <alias>NF2000tropmam4l32</alias>
```
Can we also use an uppercase `L` here?
```suggestion
    <alias>NF2000tropmam4L32</alias>
```

gold2718 commented on cime_config/config_compsets.xml:
```diff
@@ -41,24 +41,53 @@
+    <alias>NF2000tropmam4l32</alias>
+    <lname>2000_CAM%DEV%NORESM%TROPMAM4_CLM50%SP_CICE%PRES_DOCN%DOM_MOSART_SGLC_SWAV</lname>
+  </compset>
+
+  <compset>
+    <alias>NF2000climo</alias>
+    <lname>2000_CAM60%NORESM_CLM50%SP_CICE%PRES_DOCN%DOM_MOSART_SGLC_SWAV</lname>
+  </compset>
```
I don't think this will work right now (no support for CAM6 with NorESM physics).

gold2718 commented on cime_config/config_compsets.xml:
```diff
@@ -41,24 +41,53 @@
     <alias>NF1850</alias>
     <lname>1850_CAM60%NORESM_CLM51%FATES-SP_CICE%PRES_DOCN%DOM_MOSART_SGLC_SWAV</lname>
   </compset>
+  <compset>
+    <alias>NF1850oslo</alias>
+    <lname>1850_CAM60%NORESM%CAMoslo_CLM51%FATES-SP_CICE%PRES_DOCN%DOM_MOSART_SGLC_SWAV</lname>
+  </compset>
```
Ditto, no CAM6 support.

gold2718 commented on bld/namelist_files/use_cases/2000_cam6_noresm.xml:
```diff
@@ -0,0 +1,163 @@
+<?xml version="1.0"?>
```
Are the differences between this file and the NorESM2 release version necessary?

I think we should also restore the other CAM-Nor use case files:
```
bld/namelist_files/use_cases/2000_cam6_noclb_oslo.xml
bld/namelist_files/use_cases/2000_cam6_noclb_oslonudge.xml
bld/namelist_files/use_cases/2000_cam6_noresm_frc2.xml
bld/namelist_files/use_cases/2000_cam6_oslo.xml
bld/namelist_files/use_cases/2000_cam6_oslonudge.xml
```

gold2718 reviewed (commented):
This is a partial review. I think the compsets are okay (after my suggested changes) and the test set is a good initial look into the new code (we'll add others later).

mvertens committed:
use cases updated

mvertens committed:
updates to get failed tests working

Commit:
Add new script to check for differences in the energy
   diagnostics of two log files.
Move hard-coded use_hetfrz_classnuc to namelist_defaults
Replace ERS test with PEM test

mvertens committed:
removed hardwiring of use_hetfrz_classnuc

Commit:
Update externals to match new tags for CIME, CAM-Nor phys, and Oslo Aero

mvertens committed:
Merge branch 'feature/oslo_aero_integrate' into oslo_aero_integrate

mvertens committed:
Merge pull request #1 from gold2718/oslo_aero_integrate

Oslo aero integrate

Update Externals.cfg with CIME tag, cime6.0.173_noresm_v0
Update Externals_CAM.cfg with camnor_phys tag, camnor_noresm_v0.0.5
Update Externals_CAM.cfg with OSLO_AERO tag, noresm_oslo_aero_v0
More use_hetfrz_classnuc default to namelist_defaults_cam.xml
Introduce script to check for any changes in energy diagnostics

gold2718 reviewed (dismissed):
These comments will be addressed in a future update.

gold2718 committed:
Merge branch 'noresm_develop' into feature/oslo_aero_integrate

gold2718 reviewed (approved):
Resolved conflict in Externals_CAM.cfg
